Peripherals/CAN/CAN.c

```diff
@@ -1,180 +1,233 @@
+
+	// Configure PB8 for CAN RX and PB9 for CAN TX
+	RCC->APB2ENR |= RCC_APB2ENR_IOPBEN; // Enable GPIOB clock
+	GPIOB->CRH &= ~(GPIO_CRH_CNF9 | GPIO_CRH_MODE9 | GPIO_CRH_CNF8 | GPIO_CRH_MODE8);
+	GPIOB->CRH |= (0x2 << GPIO_CRH_CNF9_Pos) | (0x3 << GPIO_CRH_MODE9_Pos) | // Alternate function output push-pull, fastest speed
+				  (0x2 << GPIO_CRH_CNF8_Pos) | (0x0 << GPIO_CRH_MODE8_Pos);	 // Input
+
+	RCC->APB1ENR |= RCC_APB1ENR_CAN1EN; // Enable CAN clock
+
+	// Enable auto bus management, disable automatic retransmission, and set CAN to initialization mode
+	// Note: SLEEP and DBF (debug freeze) bits are also cleared
+	CAN1->MCR = CAN_MCR_ABOM | CAN_MCR_NART | CAN_MCR_INRQ;
+
+	// Set bit rates
+	CAN1->BTR &= ~(CAN_BTR_SJW | CAN_BTR_TS2 | CAN_BTR_TS1 | CAN_BTR_BRP);
+	CAN1->BTR |= (((can_configs[bitrate].TS2 - 1) & 0x7) << CAN_BTR_TS2_Pos) |	// Time segment 2
+				 (((can_configs[bitrate].TS1 - 1) & 0xF) << CAN_BTR_TS1_Pos) |	// Time segment 1
+				 (((can_configs[bitrate].BRP - 1) & 0x1FF) << CAN_BTR_BRP_Pos); // Baud rate prescaler
+
+	// Configure  default values
+	CAN1->FMR |= CAN_FMR_FINIT; // Set to filter initialization mode
+	CAN1->FMR |= 0x1C << CAN_FMR_CAN2SB_Pos; // Assign all filters to CAN1
+	CAN1->FA1R &= ~CAN_FA1R_FACT0;			 // Deactivate filter 0
+	CAN1->FS1R |= CAN_FS1R_FSC0;			 // Set filter 0 to single 32-bit scale configuration
+
+	// Set filter bank 0 filters to 0
+	CAN1->sFilterRegister[0].FR1 = 0x0UL;
+	CAN1->sFilterRegister[0].FR2 = 0x0UL;
+
+	CAN1->FM1R &= ~CAN_FM1R_FBM0; // Set filter to mask mode
+
+	CAN1->FFA1R &= ~CAN_FFA1R_FFA0; // Apply filter to FIFO 0
+	CAN1->FA1R |= CAN_FA1R_FACT0;	// Activate filter 0
+
+	CAN1->FMR &= ~CAN_FMR_FINIT; // Exit filter initialization mode
+	CAN1->MCR &= ~CAN_MCR_INRQ;	 // Set CAN to normal mode
+
+	while (CAN1->MSR & 0x1UL); // Wait for CAN to enter normal mode
+}
+
+/**
+ * Configures the next available CAN receive filter to pass the given ID
+ * 
+ * @param id The ID to accept
+ * 
+ * @returns 0 if successful, -1 if all of the filters are in use
  */
- void CANInit(enum BITRATE bitrate)
- {
-	RCC->APB1ENR |= 0x2000000UL;  	// Enable CAN clock 
-	RCC->APB2ENR |= 0x1UL;			// Enable AFIO clock
-	AFIO->MAPR   &= 0xFFFF9FFF;   	// reset CAN remap
-	AFIO->MAPR   |= 0x00004000;   	//  et CAN remap, use PB8, PB9
- 
-	RCC->APB2ENR |= 0x8UL;			// Enable GPIOB clock
-	GPIOB->CRH   &= ~(0xFFUL);
-	GPIOB->CRH   |= 0xB8UL;			// Configure PB8 and PB9
-	GPIOB->ODR |= 0x1UL << 8;
-  
-	CAN1->MCR = 0x51UL;			    // Set CAN to initialization mode
-	 
-	// Set bit rates 
-	CAN1->BTR &= ~(((0x03) << 24) | ((0x07) << 20) | ((0x0F) << 16) | (0x1FF)); 
-	CAN1->BTR |=  (((can_configs[bitrate].TS2-1) & 0x07) << 20) | (((can_configs[bitrate].TS1-1) & 0x0F) << 16) | ((can_configs[bitrate].BRP-1) & 0x1FF);
- 
-	// Configure Filters to default values
-	CAN1->FM1R |= 0x1C << 8;              // Assign all filters to CAN1
-	CAN1->FMR  |=   0x1UL;                // Set to filter initialization mode

```

Comment:
`CAN1->FM1R |= 0x1C << 8` is supposed to be a write to `CAN1->FMR`. Although it was done wrong, fortunately the FM1R register is locked until the next operation `CAN1->FMR |= 0x1UL;` (enter filter initialization mode). So, it just did nothing rather than what it was supposed to do. This has been corrected.